Subject: [PATCH] explorer: Add a File Manager item to the start menu.

On Android there is no host file manager next to the Wine desktop, and the
start menu only had the Start Menu folders, Control Panel and Run. Add a
File Manager entry that starts winefile.
---
diff --git a/programs/explorer/explorer.rc b/programs/explorer/explorer.rc
index 72cd9a1..90d84a9 100644
--- a/programs/explorer/explorer.rc
+++ b/programs/explorer/explorer.rc
@@ -32,6 +32,7 @@ STRINGTABLE
     IDS_PATHBOX_LABEL   "Location:"
     IDS_START_LABEL     "Start"
     IDS_RUN             "&Run..."
+    IDS_FILE_MANAGER    "&File Manager"
     IDS_EXIT_LABEL      "E&xit desktop"
     IDS_EXIT_PROMPT     "Are you sure you want to close all programs and exit the virtual desktop?"
     IDS_DESKTOP_TITLE   "Wine Desktop"
diff --git a/programs/explorer/resource.h b/programs/explorer/resource.h
index e3bf993..6249b6e 100644
--- a/programs/explorer/resource.h
+++ b/programs/explorer/resource.h
@@ -28,5 +28,6 @@
 #define IDS_EXIT_LABEL 5
 #define IDS_EXIT_PROMPT 6
 #define IDS_DESKTOP_TITLE 7
+#define IDS_FILE_MANAGER 8
 
 #endif
diff --git a/programs/explorer/startmenu.c b/programs/explorer/startmenu.c
index 6312002..36d4cd6 100644
--- a/programs/explorer/startmenu.c
+++ b/programs/explorer/startmenu.c
@@ -54,6 +54,7 @@ static struct menu_item user_startmenu;
 
 #define MENU_ID_RUN 1
 #define MENU_ID_EXIT 2
+#define MENU_ID_FILES 3
 
 static ULONG copy_pidls(struct menu_item* item, LPITEMIDLIST dest)
 {
@@ -431,6 +432,8 @@ LRESULT menu_wndproc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam)
                 exec_item(item);
             else if (mii.wID == MENU_ID_RUN)
                 run_dialog();
+            else if (mii.wID == MENU_ID_FILES)
+                ShellExecuteW(NULL, NULL, L"winefile.exe", NULL, NULL, SW_SHOWNORMAL);
             else if (mii.wID == MENU_ID_EXIT)
                 shut_down(hwnd);
 
@@ -486,6 +489,13 @@ void do_startmenu(HWND hwnd)
         AppendMenuW(root_menu.menuhandle, MF_SEPARATOR, 0, NULL);
     }
 
+    LoadStringW(NULL, IDS_FILE_MANAGER, label, ARRAY_SIZE(label));
+    mii.cbSize = sizeof(mii);
+    mii.fMask = MIIM_STRING|MIIM_ID;
+    mii.dwTypeData = label;
+    mii.wID = MENU_ID_FILES;
+    InsertMenuItemW(root_menu.menuhandle, -1, TRUE, &mii);
+
     if (SUCCEEDED(SHGetSpecialFolderLocation(NULL, CSIDL_CONTROLS, &pidl)))
         add_shell_item(&root_menu, pidl);
 
